main.c

```diff
@@ -111,6 +118,80 @@ struct tstatstruct {
 VEC_STRUCT(tstatsvec,struct tstatstruct);
 #endif
 
+typedef union {
+	u8 raw[pkprefixlen + PUBLIC_LEN + 32];
+	struct {
+		u64 prefix[4];
+		u64 key[4];
+		u64 hash[4];
+	} i;
+} pubonionunion;
+
+
+
+#define BUF_APPEND(buf,offset,src,srclen) strncpy(&buf[offset],src,srclen); offset += srclen;
+#define BUF_APPEND_CSTR(buf,offset,src) BUF_APPEND(buf,offset,src,strlen(src))
+#define BUF_APPEND_CHAR(buf,offet,c) buf[offset++] = c;
+
+static void writekeys(const char *hostname, const u8 *formated_secret, const u8 *formated_public)
+{
+	char keysbuf[KEYS_LEN];
+	size_t offset = 0;
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_generated);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_hostname);
+	BUF_APPEND(keysbuf, offset, hostname, ONION_LEN);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_secretkey);
+	char seckeybuf[SECKEY_LEN + NULLTERM_LEN];
+	base64_to(seckeybuf, formated_secret, FORMATTED_SECRET_LEN);
+	BUF_APPEND(keysbuf, offset, seckeybuf, SECKEY_LEN);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_publickey);
+	char pubkeybuf[PUBKEY_LEN + NULLTERM_LEN];
+	base64_to(pubkeybuf, formated_public, FORMATTED_PUBLIC_LEN);
+	BUF_APPEND(keysbuf, offset, pubkeybuf, PUBKEY_LEN);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_time);
+	char timebuf[TIME_LEN + NULLTERM_LEN];
+	time_t timer;
+	struct tm* tm_info;
+	time(&timer);
+	tm_info = localtime(&timer);
+	strftime(timebuf, TIME_LEN + NULLTERM_LEN, "%Y-%m-%d %H:%M:%S", tm_info);
```

Comment:
either should end with Z (`"%Y-%m-%d %H:%M:%SZ"`) (and use gmtime) or use %z (`"%Y-%m-%d %H:%M:%S%z"`) to indicate timezone. from looking around a bit, it seems like %z, while not being standard ISO C, is quite widely supported. it appears that [yaml spec for timestamp](http://yaml.org/type/timestamp.html) allows for space before 'Z' or timezone offset so may include that for better readability.

Reply:
tho no.. %z prints in format of `+hhmm`/`-hhmm` while yaml wants `[+-]hh[:mm]` so that doesn't really fit without some sort of conversion

Reply:
I've also read somewhere that `%z` in windows outputs name instead of numeric offset..
tbh just using `gmtime` + Z seems like easy way to avoid this mess, but that has less chances to be easier to understand for user.
this is all annoying.